src/dawn/SIR/SIR.h

```diff
@@ -348,9 +361,47 @@ struct Value : NonCopyable {
 
   /// @brief Set the value
   template <class T>
-  void setValue(const T& value) {
-    valueImpl_ = make_unique<ValueImpl<T>>(value);
-    type_ = TypeInfo<T>::Type;
+  typename std::enable_if<!std::is_arithmetic<T>::value>::type
+  setValue(const T& value) {
+    if (type_ == TypeKind::None) {
+      valueImpl_ = make_unique<ValueImpl<T>>(value);
+      type_ = TypeInfo<T>::Type;
+    } else {
+      if (TypeInfo<T>::Type == type_) {
```

Comment:
So for non-arithmetic types, the behaviour is different and T actually needs to be the same as `type`. For arithmetic it only needs to be convertible. This is inconsistent.